src/System.Text.Primitives/System/Text/Encoders/Utf8Decoder.cs

```diff
@@ -0,0 +1,264 @@
+            else
+            {
+                minAllowedCodePointValue = 0;
+                return 0;
+            }
+        }
+
+        /// <summary>
+        /// Returns true iff the specified code point is a UTF16 surrogate.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsSurrogate(int codePoint)
+        {
+            Debug.Assert(0 <= codePoint && codePoint <= 0x10FFFF);
+
+            // See http://www.unicode.org/versions/Unicode10.0.0/ch03.pdf, Sec. 3.8.
+            // Surrogates are in the range U+D800..U+DFFF, so we can perform a simple
+            // bit mask check to see if the requested code point is within this range.
+            return ((codePoint & 0b1111_1111_1111_1000_0000_0000) == 0b1101_1000_0000_0000);
+        }
+
+        /// <summary>
+        /// Returns true iff the specified byte is a valid UTF8 trailing byte.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsValidTrailingByte(byte value)
+        {
+            // See http://www.unicode.org/versions/Unicode10.0.0/ch03.pdf, Table 3-6.
+            return ((value & 0b1100_0000) == 0b1000_0000);
+        }
+
+        /// <summary>
+        /// Determines whether the provided data represents a well-formed UTF8 string.
+        /// </summary>
+        /// <param name="data">The data to check for well-formedness.</param>
+        /// <returns>True if the data is a well-formed UTF8 string; false otherwise.</returns>
+        public static bool IsWellFormedUtf8String(ReadOnlySpan<byte> data)
+        {
+            return (GetIndexOfFirstInvalidUtf8Byte(data) < 0);
+        }
+
+        /// <summary>
+        /// Reads a Unicode scalar value from the provided UTF8 sequence.
+        /// </summary>
+        /// <param name="data">The UTF8 sequence from which to read the scalar value.</param>
+        /// <param name="bytesConsumed">When this method returns, contains the number of bytes
+        /// from <paramref name="data"/> which were consumed while reading the scalar value.</param>
+        /// <returns>On success, a non-negative integer which represents a Unicode scalar value.
+        /// On failure, a negative integer which represents an <see cref="ErrorStatus"/> failure code.</returns>
+        /// <remarks>See http://www.unicode.org/glossary/#unicode_scalar_value for the definition of Unicode scalar value.</remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int ReadUnicodeScalarValue(ReadOnlySpan<byte> data, out int bytesConsumed)
+        {
+            return DangerousReadUnicodeScalarValue(ref data.DangerousGetPinnableReference(), data.Length, out bytesConsumed);
+        }
+
+        // Assumption: bytesAvailable > 0
+        // Assumption: bytesConsumed = 0
+        private static int ReadUnicodeScalarValueFromNonAscii(ref byte data, int bytesAvailable, ref int bytesConsumed)
```

Comment:
If bytesConsumed is effectively required to be set to 0 on input, why is this not just an out parameter?